Feat1 : Connexion, Inscription

xaviersanc opened the pull request:
# Pull Request – Ajout du Front-End : Connexion, Inscription
---

## Numéro du ticket

Ticket #2 

## Description des changements apportés

### 1.  Authentification locale

* Création d’une page d’inscription (register.html.twig) permettant l’enregistrement des utilisateurs dans la base MariaDB.
* Création d’un formulaire Symfony (RegistrationFormType) avec vérification et hachage des mots de passe.
* Mise en place de la validation côté serveur (champs obligatoires, contraintes de longueur, email unique).
* Contrôleur ListeVMController gérant l’affichage et le traitement du formulaire.

### 2.  Connexion utilisateur
* Page de connexion (connection.html.twig) reliée au système de sécurité de Symfony.
* Vérification des identifiants et affichage des erreurs en cas d’échec d’authentification.
* Gestion de session utilisateur et option “remember me” pour conserver la connexion.
* Protection des routes : redirection automatique vers la page de connexion si l’utilisateur n’est pas connecté.

### 3. Connexion via GitHub (OAuth2)

* Intégration du bundle knpuniversity/oauth2-client-bundle et du client GitHub.
* Ajout du contrôleur GitHubController et de l’authenticator GitHubAuthenticator.
* Authentification via OAuth2 permettant à un utilisateur GitHub de se connecter ou de créer un compte automatiquement.
* Configuration sécurisée dans .env via les variables GITHUB_CLIENT_ID et GITHUB_CLIENT_SECRET.
* Redirection automatique vers le tableau de bord après validation.

### 4. Tableau de bord (interface simulée)

* Création d’une page d’index (liste_vm/index.html.twig) représentant une simulation d’un dashboard.
* Présentation des machines virtuelles sous forme de cartes et de tableau compact.
* Affichage d’un état visuel pour chaque VM (en ligne, hors ligne, en maintenance).
* Actions non fonctionnelles pour cette version (boutons désactivés, interface visuelle uniquement).
* Accès restreint : page uniquement accessible après connexion.

### 5. Barre de navigation réutilisable

* Mise en place d’un partial Twig (partials/_navbar.html.twig) intégré sur toutes les pages.
* Affichage dynamique du nom de l’utilisateur connecté (initiale du prénom + nom complet).
* Menu déroulant proposant l’accès à l’espace membre et la déconnexion.
* Si aucun utilisateur connecté : bouton “Se connecter” redirigeant vers la page de connexion.
---

## Améliorations techniques

Ajout de la configuration complète de sécurité (security.yaml), avec :
* firewall main (form_login + remember_me),
* provider Doctrine (User2),
* logout sécurisé,
* authenticator GitHub.
* Validation des entités via contraintes Symfony (Assert@NotBlank, UniqueEntity).
* Intégration d’un design unifié basé sur Pico.css.
* Navigation fluide entre toutes les pages (accueil, connexion, inscription, tableau de bord).

---

## Liste de vérification des tests

## Liste de vérification avant de demander une relecture (chaîne magique, indentation, commentaires/documentation...)

- [ ] J’ai détaillé les tests à effectuer pour ma fonctionnalité/correction afin d’éviter d’éventuelles régressions (à préciser dans **Liste de vérification des tests**)
- [x] J’ai relu mon propre code
- [x] J’ai commenté mon code, en particulier dans les parties difficiles à comprendre
- [ ] J’ai apporté les modifications correspondantes à la documentation (documentation API, etc.) — à préciser dans la **Description** pour la version cible
- [ ] S’il s’agit d’une fonctionnalité conséquente, j’ai ajouté des tests approfondis
- [ ] J’ai ajouté des tests qui prouvent que ma correction est efficace ou que ma fonctionnalité fonctionne correctement
- [ ] Toutes les modifications dépendantes ont été intégrées à ce projet (à préciser dans la **Description**)

Commit:
feat : create controller "listVM"

xaviersanc committed:
Feat2 : Ajout page de connexion

xaviersanc committed:
Feet1_apparence_connexion

xaviersanc committed:
Feat2 : inscription simplifiée

xaviersanc committed:
feat2_termine

xaviersanc committed:
feat2_final

ClemLcs commented:
@xaviersanc j'ai mis à jour la description de ta MR selon un template commun.

ClemLcs commented on .env.dev:
```diff
@@ -2,3 +2,8 @@
 ###< symfony/framework-bundle ###
+
+###> github oauth ###
+GITHUB_CLIENT_ID=Ov23liR52dAXdv6kTogZ
+GITHUB_CLIENT_SECRET=7c0bd8046862a6ad254abdad1a2d4f8628f51ab1
```
Attention Xavier si c'est des vrais identifiants ils ne doivent jamais être commit

ClemLcs commented on config/packages/security.yaml:
```diff
@@ -1,39 +1,41 @@
+
+  providers:
+    app_user_provider:
+      entity:
+        class: App\Entity\User2
```
Je ne suis pas fan du nom de cette classe. “User2”, ça signifie quoi ?

ClemLcs committed:
Delete migrations/Version20251015091936.php

Il n’est pas nécessaire de committer les fichiers de migration, car on peut les générer individuellement

ClemLcs committed:
Delete migrations/Version20251010084650.php

ClemLcs committed:
Delete migrations/Version20251011094559.php

Il n’est pas nécessaire de committer les fichiers de migration, car on peut les générer individuellement

ClemLcs committed:
Delete migrations/Version20251015084128.php

Il n’est pas nécessaire de committer les fichiers de migration, car on peut les générer individuellement

ClemLcs committed:
Delete migrations/Version20251015085814.php

Il n’est pas nécessaire de committer les fichiers de migration, car on peut les générer individuellement

ClemLcs committed:
Delete migrations/Version20251015091821.php

Il n’est pas nécessaire de committer les fichiers de migration, car on peut les générer individuellement

ClemLcs commented on src/Controller/ListeVMController.php:
```diff
@@ -0,0 +1,90 @@
+    }
+
+
+    #[Route('/register', name: 'app_register')]
```
Si tes routes sont en Français plus haut c'est plus logique qu'elle soit en Français aussi ici

ClemLcs commented on src/Controller/ListeVMController.php:
```diff
@@ -0,0 +1,90 @@
+        ]);
+    }
+
+    #[Route('/forgeten_password', name: 'app_forgeten_password')]
```
Idem

ClemLcs commented on src/Controller/ListeVMController.php:
```diff
@@ -0,0 +1,90 @@
+        ]);
+    }
+
+    #[Route('/logout', name: 'app_logout', methods: ['GET'])]
```
Idem

ClemLcs commented on src/Controller/ListeVMController.php:
```diff
@@ -0,0 +1,90 @@
+    #[Route('/logout', name: 'app_logout', methods: ['GET'])]
+    public function logout(): void
+    {
+        // Intercepté par le firewall. Ne sera jamais exécuté.
```
Comment ça l'utilisateur ne peut pas se déconnecter ?

ClemLcs commented on src/Controller/ListeVMController.php:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    
+    #[Route('/liste/v/m', name: 'app_liste_v_m')]
```
/liste/vm non ?

ClemLcs commented on src/Entity/User.php:
```diff
@@ -0,0 +1,54 @@
+    #[ORM\Id]
+    #[ORM\GeneratedValue]
+    #[ORM\Column]
+    private ?int $id = null;
```
Un ID ne peut jamais être null. Donc ton auttribut aussi
```suggestion
    private int $id = 1;
```

ClemLcs commented on src/Entity/User.php:
```diff
@@ -0,0 +1,54 @@
+    #[Assert\NotBlank]
+    #[Assert\Email]
+    #[Assert\Length(max: 255)]
+    private ?string $email = null;
```
Item si l'email ne peut pas être null à cause du #[Assert\NotBlank] alors
```suggestion
    private string $email = "";
```

ClemLcs commented on src/Entity/User.php:
```diff
@@ -0,0 +1,54 @@
+
+    #[ORM\Column(length: 255)]
+    #[Assert\NotBlank]
+    private ?string $password = null;
```
Idem
```suggestion
    private string $password = "";
```

ClemLcs commented on src/Entity/User.php:
```diff
@@ -0,0 +1,54 @@
+
+    // Get / Set 
+
+    public function getId(): ?int { return $this->id; }
+
+    public function getFirstName(): ?string { return $this->firstName; }
+    public function setFirstName(string $firstName): self { $this->firstName = $firstName; return $this; }
+
+    public function getLastName(): ?string { return $this->lastName; }
+    public function setLastName(string $lastName): self { $this->lastName = $lastName; return $this; }
+
+    public function getEmail(): ?string { return $this->email; }
+    public function setEmail(string $email): self { $this->email = $email; return $this; }
+
+    public function getPassword(): ?string { return $this->password; }
+    public function setPassword(string $password): self { $this->password = $password; return $this; }
```
Mettre à jour les getter et setter du coup

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+#[ORM\Entity(repositoryClass: User2Repository::class)]
+#[ORM\UniqueConstraint(name: 'UNIQ_IDENTIFIER_EMAIL', fields: ['email'])]
+#[UniqueEntity(fields: ['email'], message: 'There is already an account with this email')]
+class User2 implements UserInterface, PasswordAuthenticatedUserInterface
```
A quoi fait référence cette entity ? User2 ?

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+    #[ORM\Id]
+    #[ORM\GeneratedValue]
+    #[ORM\Column]
+    private ?int $id = null;
```
```suggestion
    private int $id = 0;
```

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+    #[ORM\Column(length: 180)]
+    #[Assert\NotBlank]
+    #[Assert\Email]
+    private ?string $email = null;
```
```suggestion
    private string $email = "";
```

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+
+    #[ORM\Column(length: 180)]
+    #[Assert\NotBlank]
+    private ?string $firstName = null;
```
```suggestion
    private string $firstName = "";
```

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+
+    #[ORM\Column(length: 180)]
+    #[Assert\NotBlank]
+    private ?string $lastName = null;
```
```suggestion
    private string $lastName = "";
```

ClemLcs commented on src/Entity/User2.php:
```diff
@@ -0,0 +1,119 @@
+    private array $roles = [];
+
+    #[ORM\Column]
+    private ?string $password = null;
```
Pourquoi a t'on besoin de stocker le password ?

ClemLcs commented on src/Repository/User2Repository.php:
```diff
@@ -0,0 +1,60 @@
+        $this->getEntityManager()->flush();
+    }
+
+    //    /**
+    //     * @return User2[] Returns an array of User2 objects
+    //     */
+    //    public function findByExampleField($value): array
+    //    {
+    //        return $this->createQueryBuilder('u')
+    //            ->andWhere('u.exampleField = :val')
+    //            ->setParameter('val', $value)
+    //            ->orderBy('u.id', 'ASC')
+    //            ->setMaxResults(10)
+    //            ->getQuery()
+    //            ->getResult()
+    //        ;
+    //    }
+
+    //    public function findOneBySomeField($value): ?User2
+    //    {
+    //        return $this->createQueryBuilder('u')
+    //            ->andWhere('u.exampleField = :val')
+    //            ->setParameter('val', $value)
+    //            ->getQuery()
+    //            ->getOneOrNullResult()
+    //        ;
+    //    }
```
Si ce code est inutile autant le supprimer
```suggestion
```

ClemLcs commented on src/Repository/UserRepository.php:
```diff
@@ -0,0 +1,43 @@
+        parent::__construct($registry, User::class);
+    }
+
+    //    /**
+    //     * @return User[] Returns an array of User objects
+    //     */
+    //    public function findByExampleField($value): array
+    //    {
+    //        return $this->createQueryBuilder('u')
+    //            ->andWhere('u.exampleField = :val')
+    //            ->setParameter('val', $value)
+    //            ->orderBy('u.id', 'ASC')
+    //            ->setMaxResults(10)
+    //            ->getQuery()
+    //            ->getResult()
+    //        ;
+    //    }
+
+    //    public function findOneBySomeField($value): ?User
+    //    {
+    //        return $this->createQueryBuilder('u')
+    //            ->andWhere('u.exampleField = :val')
+    //            ->setParameter('val', $value)
+    //            ->getQuery()
+    //            ->getOneOrNullResult()
+    //        ;
+    //    }
```
```suggestion
```

ClemLcs commented on src/Security/EmailVerifier.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        $signatureComponents = $this->verifyEmailHelper->generateSignature(
+            $verifyEmailRouteName,
+            (string) $user->getId(),
```
Pourquoi castes-tu ton id en string ?

ClemLcs commented on src/Security/EmailVerifier.php:
```diff
@@ -0,0 +1,52 @@
+        $signatureComponents = $this->verifyEmailHelper->generateSignature(
+            $verifyEmailRouteName,
+            (string) $user->getId(),
+            (string) $user->getEmail()
```
Idem ici

ClemLcs commented on src/Security/GitHubAuthenticator.php:
```diff
@@ -0,0 +1,66 @@
+
+                if (!$user) {
+                    $user = new User2();
+                    $user->setEmail($githubUser->getEmail() ?? $githubUser->getId().'@github.local');
```
Que signifie les ??

ClemLcs commented on src/Security/GitHubAuthenticator.php:
```diff
@@ -0,0 +1,66 @@
+                    $user->setEmail($githubUser->getEmail() ?? $githubUser->getId().'@github.local');
+                    $user->setFirstName($githubUser->getNickname() ?? 'Git');
+                    $user->setLastName('Hub');
+                    $user->setPassword(''); // inutile pour OAuth
```
Si inutile alors pourquoi avec cette attribut ?

ClemLcs commented on structure.txt:
A quoi sert ce fichier ?

ClemLcs commented on templates/ForgetenPassword.html.twig:
```diff
@@ -0,0 +1,132 @@
+        href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.1.1/css/pico.min.css" />
+
+    <!-- Personnalisation -->
+  <style>
```
Si tu fais un css custom autant le mettre dans un fichier style.css car sinon c'est moche

ClemLcs commented on templates/ForgetenPassword.html.twig:
```diff
@@ -0,0 +1,132 @@
+      <details class="dropdown">
+        <summary role="button" class="secondary">Accueil</summary>
+        <ul class="dropdown-menu">
+          <li><a href="/liste/v/m">Liste des VM</a></li>
```
Est-ce qu'il ne serait pas plus judicieux d'utiliser le nom de route plutôt ?

ClemLcs commented on templates/connection.html.twig:
```diff
@@ -0,0 +1,96 @@
+  <title>VM - Se connecter</title>
+  <meta name="description" content="Il faut se connecter, c'est tout" />
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.1.1/css/pico.min.css" />
+  <style>
+    header { position: sticky; top: 0; z-index: 10; background-color: rgba(240,240,240,.7); backdrop-filter: blur(4px); border-bottom: 1px solid rgba(0,0,0,.05); }
+    header nav { max-width: 1000px; margin: 0 auto; padding: .5rem 1rem; }
+    body { min-height: 100dvh; display: flex; flex-direction: column; }
+    main.page { flex: 1; display: flex; align-items: center; justify-content: center; padding: 2rem 1rem; }
+    .content { display: flex; flex-direction: column; gap: 1.25rem; width: min(680px, 100%); }
+    .block { padding: 1rem 1.25rem; border-radius: .5rem; background: var(--pico-card-background-color); box-shadow: var(--pico-card-box-shadow); }
+    .block h2, .block h3 { margin: 0 0 .5rem 0; }
+    form .actions { display: flex; gap: .75rem; flex-wrap: wrap; }
+    @media (min-width: 960px) { .content { gap: 1.5rem; } .block { padding: 1.25rem 1.5rem; } }
+  </style>
```
Idem

ClemLcs commented on templates/home.html.twig:
```diff
@@ -0,0 +1,38 @@
+  <title>VM - Accueil</title>
+  <meta name="description" content="Description VM" />
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.1.1/css/pico.min.css" />
+  <style>
+    body { min-height: 100dvh; display: flex; flex-direction: column; }
+    main { flex: 1; display: flex; align-items: center; justify-content: center; padding: 2rem 1rem; }
+    section { text-align: center; display: flex; flex-direction: column; gap: 1.5rem; width: min(480px, 100%); }
+    .nav-links a {
+      display: block; margin: .5rem 0; padding: .75rem 1rem; border-radius: .5rem; text-decoration: none; font-weight: 500;
+      background: var(--pico-card-background-color); box-shadow: var(--pico-card-box-shadow);
+    }
+    .nav-links a:hover { background: var(--pico-primary); color: #fff; }
+  </style>
```
Idem

ClemLcs commented on templates/liste_vm/index.html.twig:
```diff
@@ -0,0 +1,247 @@
+  <meta name="viewport" content="width=device-width,initial-scale=1" />
+  <title>VM — Tableau de bord</title>
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.1.1/css/pico.min.css" />
+  <style>
```
Idem

ClemLcs commented on templates/partials/_navbar.html.twig:
```diff
@@ -0,0 +1,44 @@
+<header>
```
Si tu as un fichier navbar, pourquoi ne pas charger le fichier CSS ici ?

ClemLcs commented on templates/register.html.twig:
```diff
@@ -0,0 +1,152 @@
+  <link rel="stylesheet"
+        href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.1.1/css/pico.min.css" />
+
+  <style>
```
A mettre dans un fichier séparer

ClemLcs commented on templates/register.html.twig:
```diff
@@ -0,0 +1,152 @@
+      <details class="dropdown">
+        <summary role="button" class="secondary">Accueil</summary>
+        <ul class="dropdown-menu">
+          <li><a href="/liste/v/m">Liste des VM</a></li>
```
Utiliser les nom de route pour générer automatiquement les uri

ClemLcs commented on templates/register.html.twig:
```diff
@@ -0,0 +1,152 @@
+        <summary role="button" class="secondary">Accueil</summary>
+        <ul class="dropdown-menu">
+          <li><a href="/liste/v/m">Liste des VM</a></li>
+          <li><a href="/connection">Se connecter</a></li>
```
Utiliser les nom de route pour générer automatiquement les uri
